docs/design/repair-with-hashes.md

```diff
@@ -0,0 +1,63 @@
+3. The piece private key is included in the create segment, download segment or delete a segment.
+4. The uplink uses piece private key to sign the orders.
+5. The storage node verifies order limit signature as usual and then uses the piece public key to verify the order and piece hash.
+6. Storage node sends both the order limit and order to the satellite during settlement.
+7. Storage node sends data, order limit and piece hash to the satellite, if it needs to prove data-consistency to the satellite.
+
+Thus the uplink-signed `uplink_piece_hash` is already stored in the storage node's `pieceinfo_` table as a serialized protocol buffer and includes the `PieceID`, `Hash`, and uplink `Signature`. The `pieceinfo_` table also includes the `order_limit` which contains the `PiecePublicKey` necessary for validating the `PieceHash`, as well as the `SatelliteSignature` to validate the `OrderLimit` itself.
+
+## Design
+
+When a satellite is download pieces for repair, the storage node should also include the `PieceHash` and `OrderLimit` to the satellite along with each piece.
+
+The satellite should attempt to download exactly the Reed-Solomon minimum number of pieces during a repair, storing them in memory. It should validate the `OrderLimit` using its own private key.  It should validate that the `PieceHash` `Signature` corresponds to the `PublicKey` contained in the `OrderLimit`.  The satellite should calculate the size and hash of the piece as it downloads them, and it should be equal to the `PieceHash` `Hash` and `PieceSize`. If a piece fails any of these checks, an error should be logged and a different piece should be downloaded.  Once all pieces are downloaded, the `PiecePublicKey` of all pieces should be compared and equal.  If they are not equal, an error should be logged and more pieces should be downloaded until a minimum number of pieces share the same `PiecePublicKey`.  When enough pieces have been downloaded and validated, Reed-Solomon decoding should be used to reassemble the segment, and repair should continue as previously designed.
+
+Failed hash, signature, or public keys checks may indicate cheating. Therefore audit must implement this same piece hash check logic. Changes to audit and the reporting of cheaters are otherwise outside the scope of this document.  
+
+## Rationale
+
+Both download and Reed Solomon decoding create more load for a satellite than hashing, so hashing should be preferred over download even a single more piece.
+
+Downloading for repair is significantly different enough from streaming as to warrant a new, custom implementation.
+
+Using only the minimum number of pieces means that Reed-Solomon does not act as a check during repair. Hence hashing is used instead. While and uplink could potentially send signed bogus data to a storage node, the storage node would not be penalized by these actions. This requires that Audit implements a similar piece hash check instead of relying solely on Reed-Solomon encoding.
```

Comment:
While *an uplink

Also, does a ticket already exist for implementing piece hash checks in the audit system? We could make one that's like- make a mini design doc for piece hash checking for the audit system

Reply:
yes!  https://storjlabs.atlassian.net/browse/V3-1322  It turns out that this ticket came out of design / document and this fact was never explained to us.  https://docs.google.com/document/d/1Vcu7lLPSMa9OTemGxdlLbhptGqKu1FyY4ltNHczP2lw/edit  :sigh:  I discovered this too late in the process for it to be of much use.